File: src/main/java/com/mparticle/kits/GoogleAnalyticsFirebaseKit.java
package com.mparticle.kits;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.mparticle.MPEvent;
import com.mparticle.MParticle;
import com.mparticle.commerce.CommerceEvent;
import com.mparticle.commerce.Product;
import com.mparticle.commerce.TransactionAttributes;
import com.mparticle.consent.ConsentState;
import com.mparticle.identity.MParticleUser;
import com.mparticle.internal.Logger;
import com.mparticle.internal.MPUtility;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class GoogleAnalyticsFirebaseKit extends KitIntegration implements KitIntegration.EventListener, KitIntegration.IdentityListener, KitIntegration.CommerceListener, KitIntegration.UserAttributeListener {
    final static String USER_ID_FIELD_KEY = "userIdField";
    final static String USER_ID_CUSTOMER_ID_VALUE = "customerId";
    final static String USER_ID_EMAIL_VALUE = "email";
    final static String USER_ID_MPID_VALUE = "mpid";

    final static String CF_GA4COMMERCE_EVENT_TYPE = "GA4.CommerceEventType";
    final static String CF_GA4_PAYMENT_TYPE = "GA4.PaymentType";
    final static String CF_GA4_SHIPPING_TIER = "GA4.ShippingTier";

    private static String[] forbiddenPrefixes = new String[]{"google_", "firebase_", "ga_"};
    private static int eventMaxLength = 40;
    private static int userAttributeMaxLength = 24;

    private static int eventValMaxLength = 100;
    private static int userAttributeValMaxLength = 36;

    @Override
    public String getName() {
        return "Google Analytics for Firebase";
    }

    @Override
    protected List<ReportingMessage> onKitCreate(Map<String, String> map, Context context) throws IllegalArgumentException {
        Logger.info(getName() + " Kit relies on a functioning instance of Firebase Analytics. If your Firebase Analytics instance is not configured properly, this Kit will not work");
        return null;
    }

    @Override
    public List<ReportingMessage> setOptOut(boolean b) {
        return null;
    }

    @Override
    public List<ReportingMessage> leaveBreadcrumb(String s) {
        return null;
    }

    @Override
    public List<ReportingMessage> logError(String s, Map<String, String> map) {
        return null;
    }

    @Override
    public List<ReportingMessage> logException(Exception e, Map<String, String> map, String s) {
        return null;
    }

    @Override
    public List<ReportingMessage> logEvent(MPEvent mpEvent) {
        FirebaseAnalytics.getInstance(getContext())
                .logEvent(getFirebaseEventName(mpEvent), toBundle(mpEvent.getInfo()));

        return Collections.singletonList(ReportingMessage.fromEvent(this, mpEvent));
    }

    @Override
    public List<ReportingMessage> logScreen(String s, Map<String, String> map) {
        Activity activity = getCurrentActivity().get();
        if (activity != null) {
            FirebaseAnalytics.getInstance(getContext()).setCurrentScreen(activity, standardizeName(s, true), null);
            return Collections.singletonList(new ReportingMessage(this, ReportingMessage.MessageType.SCREEN_VIEW, System.currentTimeMillis(), null));
        }
        return null;
    }

    @Override
    public List<ReportingMessage> logLtvIncrease(BigDecimal bigDecimal, BigDecimal bigDecimal1, String s, Map<String, String> map) {
        return null;
    }

    @Override
    public List<ReportingMessage> logEvent(CommerceEvent commerceEvent) {
        FirebaseAnalytics instance = FirebaseAnalytics.getInstance(getContext());
        String eventName;
        if (commerceEvent == null || commerceEvent.getProductAction() == null) {
            return null;
        }
        Bundle bundle = getCommerceEventBundle(commerceEvent)
        .getBundle();
        switch (commerceEvent.getProductAction()) {
            case Product.ADD_TO_CART:
                eventName = FirebaseAnalytics.Event.ADD_TO_CART;
                break;
            case Product.ADD_TO_WISHLIST:
                eventName = FirebaseAnalytics.Event.ADD_TO_WISHLIST;
                break;
            case Product.CHECKOUT:
                eventName = FirebaseAnalytics.Event.BEGIN_CHECKOUT;
                break;
            case Product.PURCHASE:
                eventName = FirebaseAnalytics.Event.PURCHASE;
                break;
            case Product.REFUND:
                eventName = FirebaseAnalytics.Event.REFUND;
                break;
            case Product.REMOVE_FROM_CART:
                eventName = FirebaseAnalytics.Event.REMOVE_FROM_CART;
                break;
            case Product.CLICK:
                eventName = FirebaseAnalytics.Event.SELECT_CONTENT;
                break;
            case Product.CHECKOUT_OPTION:
                Map<String, List<String>> customFlags = commerceEvent.getCustomFlags();
                if (customFlags != null && customFlags.containsKey(CF_GA4COMMERCE_EVENT_TYPE)) {
                    List<String> commerceEventTypes = customFlags.get(CF_GA4COMMERCE_EVENT_TYPE);
                    if (!commerceEventTypes.isEmpty()) {
                        String commerceEventType = commerceEventTypes.get(0);
                        if (commerceEventType.equals(FirebaseAnalytics.Event.ADD_SHIPPING_INFO.toString())) {
                            eventName = FirebaseAnalytics.Event.ADD_SHIPPING_INFO;
                        } else if (commerceEventType.equals(FirebaseAnalytics.Event.ADD_PAYMENT_INFO.toString())) {
                            eventName = FirebaseAnalytics.Event.ADD_PAYMENT_INFO;
                        } else {
                            Logger.warning("You used an unsupported value for the custom flag 'GA4.CommerceEventType'. Please review the mParticle documentation. The event will be sent to Firebase with the deprecated SET_CHECKOUT_OPTION event type.");
                            eventName = FirebaseAnalytics.Event.SET_CHECKOUT_OPTION;
                        }
                    } else {
                        Logger.warning("Setting a CHECKOUT_OPTION now requires a custom flag of 'GA4.CommerceEventType'. Please review the mParticle documentation.  The event will be sent to Firebase with the deprecated SET_CHECKOUT_OPTION event type.");
                        eventName = FirebaseAnalytics.Event.SET_CHECKOUT_OPTION;
                    }
                } else {
                    Logger.warning("Setting a CHECKOUT_OPTION now requires a custom flag of 'GA4.CommerceEventType'. Please review the mParticle documentation.  The event will be sent to Firebase with the deprecated SET_CHECKOUT_OPTION event type.");
                    eventName = FirebaseAnalytics.Event.SET_CHECKOUT_OPTION;
                }

                break;
            case Product.DETAIL:
                eventName = FirebaseAnalytics.Event.VIEW_ITEM;
                break;
            default:
                return null;
        }

        instance.logEvent(eventName, bundle);
        return Collections.singletonList(ReportingMessage.fromEvent(this, commerceEvent));
    }

    @Override
    public void onIdentifyCompleted(MParticleUser mParticleUser, FilteredIdentityApiRequest filteredIdentityApiRequest) {
        setUserId(mParticleUser);
    }

    @Override
    public void onLoginCompleted(MParticleUser mParticleUser, FilteredIdentityApiRequest filteredIdentityApiRequest) {
        setUserId(mParticleUser);
    }

    @Override
    public void onLogoutCompleted(MParticleUser mParticleUser, FilteredIdentityApiRequest filteredIdentityApiRequest) {
        setUserId(mParticleUser);
    }

    @Override
    public void onModifyCompleted(MParticleUser mParticleUser, FilteredIdentityApiRequest filteredIdentityApiRequest) {
        setUserId(mParticleUser);
    }

    @Override
    public void onUserIdentified(MParticleUser mParticleUser) {

    }

    private void setUserId(MParticleUser user) {
        String userId = null;
        if (USER_ID_CUSTOMER_ID_VALUE.equalsIgnoreCase(getSettings().get(USER_ID_FIELD_KEY))) {
            userId = user.getUserIdentities().get(MParticle.IdentityType.CustomerId);
        } else if (USER_ID_EMAIL_VALUE.equalsIgnoreCase(getSettings().get(USER_ID_FIELD_KEY))) {
            userId = user.getUserIdentities().get(MParticle.IdentityType.Email);
        } else if (user != null && USER_ID_MPID_VALUE.equalsIgnoreCase(getSettings().get(USER_ID_FIELD_KEY))) {
            userId = Long.toString(user.getId());
        }
        if (!KitUtils.isEmpty(userId)) {
            FirebaseAnalytics.getInstance(getContext()).setUserId(userId);
        }
    }

    String getFirebaseEventName(MPEvent event) {
        switch (event.getEventType()) {
            case Search:
                return FirebaseAnalytics.Event.SEARCH;
        }
        if (event.isScreenEvent()) {
            return FirebaseAnalytics.Event.VIEW_ITEM;
        }
        return standardizeName(event.getEventName(), true);
    }

    Bundle toBundle(Map<String, String> map) {
        Bundle bundle = new Bundle();
        map = standardizeAttributes(map, true);
        if (map != null) {
            for (Map.Entry<String, String> entry : map.entrySet()) {
                bundle.putString(entry.getKey(), entry.getValue());
            }
        }
        return bundle;
    }

    PickyBundle getCommerceEventBundle(CommerceEvent commerceEvent) {
        PickyBundle pickyBundle = getTransactionAttributesBundle(commerceEvent);
        String currency = commerceEvent.getCurrency();
        if (currency == null) {
            Logger.info("Currency field required by Firebase was not set, defaulting to 'USD'");
            currency = "USD";
        }

        // Google Analytics 4 introduces 2 new event types - add_shipping_info and add_payment_info
        // each of these has an extra parameter that is optional
        Map<String, List<String>> customFlags = commerceEvent.getCustomFlags();
        if (customFlags != null && customFlags.containsKey(CF_GA4COMMERCE_EVENT_TYPE)) {
            List<String> commerceEventTypeList = customFlags.get(CF_GA4COMMERCE_EVENT_TYPE);
            if (commerceEventTypeList != null && commerceEventTypeList.size() > 0) {
                String commerceEventType = commerceEventTypeList.get(0);
                if (commerceEventType.equals(FirebaseAnalytics.Event.ADD_SHIPPING_INFO.toString())) {
                    List<String> shippingTier = customFlags.get(CF_GA4_SHIPPING_TIER);
                    if (shippingTier != null && shippingTier.size() > 0) {
                        pickyBundle.putString(FirebaseAnalytics.Param.SHIPPING_TIER, shippingTier.get(0));
                    }
                } else if (commerceEventType.equals(FirebaseAnalytics.Event.ADD_PAYMENT_INFO.toString())) {
                    List<String> paymentType = customFlags.get(CF_GA4_PAYMENT_TYPE);
                    if (paymentType != null && paymentType.size() > 0) {
                        pickyBundle.putString(FirebaseAnalytics.Param.PAYMENT_TYPE, paymentType.get(0));
                    }
                }
            }
        }
        return pickyBundle
                .putString(FirebaseAnalytics.Param.CURRENCY, currency)
                .putBundleList(FirebaseAnalytics.Param.ITEMS, getProductBundles(commerceEvent))
                .putString(FirebaseAnalytics.Event.SET_CHECKOUT_OPTION, commerceEvent.getCheckoutOptions())
                .putInt(FirebaseAnalytics.Event.CHECKOUT_PROGRESS, commerceEvent.getCheckoutStep());
    }

    Bundle[] getProductBundles(CommerceEvent commerceEvent) {
        List<Product> products = commerceEvent.getProducts();
        if (products != null) {
            Bundle[] bundles = new Bundle[products.size()];
            int i = 0;
            for (Product product: products) {
                PickyBundle bundle = getBundle(product)
                        .putString(FirebaseAnalytics.Param.CURRENCY, commerceEvent.getCurrency());
                bundles[i] = bundle.getBundle();
                i++;
            }
            return bundles;
        }
        return new Bundle[0];
    }

    PickyBundle getTransactionAttributesBundle(CommerceEvent commerceEvent) {
        PickyBundle pickyBundle = new PickyBundle();
        TransactionAttributes transactionAttributes = commerceEvent.getTransactionAttributes();

        if (commerceEvent.getTransactionAttributes() == null) {
            return pickyBundle;
        }
        return pickyBundle
                .putString(FirebaseAnalytics.Param.TRANSACTION_ID, transactionAttributes.getId())
                .putDouble(FirebaseAnalytics.Param.VALUE, transactionAttributes.getRevenue())
                .putDouble(FirebaseAnalytics.Param.TAX, transactionAttributes.getTax())
                .putDouble(FirebaseAnalytics.Param.SHIPPING, transactionAttributes.getShipping())
                .putString(FirebaseAnalytics.Param.COUPON, transactionAttributes.getCouponCode());
    }

    PickyBundle getBundle(Product product) {
        return new PickyBundle()
                .putLong(FirebaseAnalytics.Param.QUANTITY, (long) product.getQuantity())
                .putString(FirebaseAnalytics.Param.ITEM_ID, product.getSku())
                .putString(FirebaseAnalytics.Param.ITEM_NAME, product.getName())
                .putString(FirebaseAnalytics.Param.ITEM_CATEGORY, product.getCategory())
                .putDouble(FirebaseAnalytics.Param.PRICE, product.getUnitPrice());
    }

    private Double getValue(CommerceEvent commerceEvent) {
        double value = 0;
        List<Product> products = commerceEvent.getProducts();
        if (products == null) {
            return null;
        }
        for (Product product: products) {
            value += product.getQuantity() * product.getUnitPrice();
        }
        return value;
    }

    @Override
    public void onIncrementUserAttribute(String key, int incrementedBy, String value, FilteredMParticleUser filteredMParticleUser) {
        FirebaseAnalytics.getInstance(getContext()).setUserProperty(standardizeName(key, false), value);
    }

    @Override
    public void onRemoveUserAttribute(String key, FilteredMParticleUser filteredMParticleUser) {
        FirebaseAnalytics.getInstance(getContext()).setUserProperty(standardizeName(key, false), null);
    }

    /**
     * We are going to ignore Lists here, since Firebase only supports String "user property" values
     */
    @Override
    public void onSetUserAttribute(String key, Object value, FilteredMParticleUser filteredMParticleUser) {
        if (value instanceof String) {
            FirebaseAnalytics.getInstance(getContext()).setUserProperty(standardizeName(key, false), standardizeValue((String)value, false));
        }
    }

    @Override
    public void onSetUserTag(String s, FilteredMParticleUser filteredMParticleUser) {

    }

    @Override
    public void onSetUserAttributeList(String s, List<String> list, FilteredMParticleUser filteredMParticleUser) {

    }

    @Override
    public void onSetAllUserAttributes(Map<String, String> userAttributes, Map<String, List<String>> userAttributeLists, FilteredMParticleUser filteredMParticleUser) {
        userAttributes = standardizeAttributes(userAttributes, false);
        for (Map.Entry<String, String> entry: userAttributes.entrySet()) {
            FirebaseAnalytics.getInstance(getContext()).setUserProperty(entry.getKey(), entry.getValue());
        }
    }

    @Override
    public boolean supportsAttributeLists() {
        return false;
    }

    @Override
    public void onConsentStateUpdated(ConsentState consentState, ConsentState consentState1, FilteredMParticleUser filteredMParticleUser) {

    }

    Map<String, String> standardizeAttributes(Map<String, String> attributes, boolean event) {
        if (attributes == null) {
            return null;
        }
        Map<String, String> attributeCopy = new HashMap<>();
        for (Map.Entry<String, String> entry: attributes.entrySet()) {
            attributeCopy.put(standardizeName(entry.getKey(), event), standardizeValue(entry.getValue(), event));
        }
        return attributeCopy;
    }

    final String standardizeValue(String value, boolean event) {
        if (value == null) {
            return value;
        }
        if (event) {
            if (value.length() > eventValMaxLength) {
                value = value.substring(0, eventValMaxLength);
            }
        } else {
            if (value.length() > userAttributeValMaxLength) {
                value = value.substring(0, userAttributeValMaxLength);
            }
        }
        return value;
    }

    final String standardizeName(String name, boolean event) {
        if (name == null) {
            return null;
        }
        name = name.replaceAll("[^a-zA-Z0-9_\\s]", "");
        name = name.replaceAll("[\\s]+", "_");

        for(String forbiddenPrefix: forbiddenPrefixes) {
            if (name.startsWith(forbiddenPrefix)) {
                name = name.replaceFirst(forbiddenPrefix, "");
            }
        }

        while(name.length() > 0 && !Character.isLetter(name.toCharArray()[0])) {
            name = name.substring(1);
        }
        if (event) {
            if (name.length() > eventMaxLength) {
                name = name.substring(0, eventMaxLength);
            }
        } else {
            if (name.length() > userAttributeMaxLength) {
                name = name.substring(0, userAttributeMaxLength);
            }
        }
        return name;
    }

    class PickyBundle {
        private Bundle bundle = new Bundle();

        PickyBundle putString(String key, String value) {
            if (value != null) {
                bundle.putString(key, value);
            }
            return this;
        }

        PickyBundle putDouble(String key, Double value) {
            if (value != null) {
                bundle.putDouble(key, value);
            }
            return this;
        }

        PickyBundle putLong(String key, Long value) {
            if (value != null) {
                bundle.putLong(key, value);
            }
            return this;
        }

        PickyBundle putInt(String key, Integer value) {
            if (value != null) {
                bundle.putInt(key, value);
            }
            return this;
        }

        PickyBundle putBundleList(String key, Bundle[] value) {
            if (value != null) {
                bundle.putParcelableArray(key, value);
            }
            return this;
        }

        Bundle getBundle() {
            return bundle;
        }
    }
}

File: src/main/kotlin/com/mparticle/kits/GoogleAnalyticsFirebaseKit.kt
package com.mparticle.kits

import android.content.Context
import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import com.mparticle.MPEvent
import com.mparticle.MParticle
import com.mparticle.MParticle.EventType
import com.mparticle.commerce.CommerceEvent
import com.mparticle.commerce.Product
import com.mparticle.consent.ConsentState
import com.mparticle.identity.MParticleUser
import com.mparticle.internal.Logger
import com.mparticle.kits.KitIntegration.CommerceListener
import com.mparticle.kits.KitIntegration.IdentityListener
import java.math.BigDecimal

class GoogleAnalyticsFirebaseKit : KitIntegration(), KitIntegration.EventListener, IdentityListener,
    CommerceListener, KitIntegration.UserAttributeListener {
    override fun getName(): String = KIT_NAME

    @Throws(IllegalArgumentException::class)
    public override fun onKitCreate(
        map: Map<String, String>,
        context: Context
    ): List<ReportingMessage> {
        Logger.info("$name Kit relies on a functioning instance of Firebase Analytics. If your Firebase Analytics instance is not configured properly, this Kit will not work")
        return emptyList()
    }

    override fun setOptOut(b: Boolean): List<ReportingMessage> = emptyList()

    override fun leaveBreadcrumb(s: String): List<ReportingMessage> = emptyList()

    override fun logError(s: String, map: Map<String, String>): List<ReportingMessage> = emptyList()

    override fun logException(
        e: Exception,
        map: Map<String, String>,
        s: String
    ): List<ReportingMessage> = emptyList()

    override fun logEvent(mpEvent: MPEvent): List<ReportingMessage> {
        getFirebaseEventName(mpEvent)?.let {
            FirebaseAnalytics.getInstance(context)
                .logEvent(it, toBundle(mpEvent.customAttributeStrings))
        }
        return listOf(ReportingMessage.fromEvent(this, mpEvent))
    }

    override fun logScreen(s: String, map: Map<String, String>): List<ReportingMessage> {
        val activity = currentActivity.get()
        if (activity != null) {
            FirebaseAnalytics.getInstance(context)
                .setCurrentScreen(activity, standardizeName(s, true), null)
            return listOf(
                ReportingMessage(
                    this,
                    ReportingMessage.MessageType.SCREEN_VIEW,
                    System.currentTimeMillis(),
                    null
                )
            )
        }
        return emptyList()
    }

    override fun logLtvIncrease(
        bigDecimal: BigDecimal,
        bigDecimal1: BigDecimal,
        s: String,
        map: Map<String, String>
    ): List<ReportingMessage> = emptyList()

    override fun logEvent(commerceEvent: CommerceEvent): List<ReportingMessage> {
        val instance = FirebaseAnalytics.getInstance(context)
        if (commerceEvent.productAction == null) {
            return emptyList()
        }
        val bundle = getCommerceEventBundle(commerceEvent)
            .bundle
        val eventName: String = when (commerceEvent.productAction) {
            Product.ADD_TO_CART -> FirebaseAnalytics.Event.ADD_TO_CART
            Product.ADD_TO_WISHLIST -> FirebaseAnalytics.Event.ADD_TO_WISHLIST
            Product.CHECKOUT -> FirebaseAnalytics.Event.BEGIN_CHECKOUT
            Product.PURCHASE -> FirebaseAnalytics.Event.PURCHASE
            Product.REFUND -> FirebaseAnalytics.Event.REFUND
            Product.REMOVE_FROM_CART -> FirebaseAnalytics.Event.REMOVE_FROM_CART
            Product.CLICK -> FirebaseAnalytics.Event.SELECT_CONTENT
            Product.CHECKOUT_OPTION ->  FirebaseAnalytics.Event.SET_CHECKOUT_OPTION
            Product.DETAIL -> FirebaseAnalytics.Event.VIEW_ITEM
            else -> return emptyList()
        }
        instance.logEvent(eventName, bundle)
        return listOf(ReportingMessage.fromEvent(this, commerceEvent))
    }

    override fun onIdentifyCompleted(
        mParticleUser: MParticleUser,
        filteredIdentityApiRequest: FilteredIdentityApiRequest
    ) {
        setUserId(mParticleUser)
    }

    override fun onLoginCompleted(
        mParticleUser: MParticleUser,
        filteredIdentityApiRequest: FilteredIdentityApiRequest
    ) {
        setUserId(mParticleUser)
    }

    override fun onLogoutCompleted(
        mParticleUser: MParticleUser,
        filteredIdentityApiRequest: FilteredIdentityApiRequest
    ) {
        setUserId(mParticleUser)
    }

    override fun onModifyCompleted(
        mParticleUser: MParticleUser,
        filteredIdentityApiRequest: FilteredIdentityApiRequest
    ) {
        setUserId(mParticleUser)
    }

    override fun onUserIdentified(mParticleUser: MParticleUser) {}
    private fun setUserId(user: MParticleUser?) {
        var userId: String? = null
        if (user != null) {
            if (USER_ID_CUSTOMER_ID_VALUE.equals(settings[USER_ID_FIELD_KEY], true)) {
                userId = user.userIdentities[MParticle.IdentityType.CustomerId]
            } else if (USER_ID_EMAIL_VALUE.equals(settings[USER_ID_FIELD_KEY], true)) {
                userId = user.userIdentities[MParticle.IdentityType.Email]
            } else if (USER_ID_MPID_VALUE.equals(settings[USER_ID_FIELD_KEY], true)) {
                userId = user.id.toString()
            }
        }
        if (!KitUtils.isEmpty(userId)) {
            FirebaseAnalytics.getInstance(context).setUserId(userId)
        }
    }

    fun getFirebaseEventName(event: MPEvent): String? {
        if (event.eventType == EventType.Search) {
            return FirebaseAnalytics.Event.SEARCH
        }
        return if (event.isScreenEvent) {
            FirebaseAnalytics.Event.VIEW_ITEM
        } else standardizeName(event.eventName, true)
    }

    fun toBundle(mapIn: Map<String, String>?): Bundle {
        var map = mapIn
        val bundle = Bundle()
        map = standardizeAttributes(map, true)
        if (map != null) {
            for ((key, value) in map) {
                bundle.putString(key, value)
            }
        }
        return bundle
    }

    fun getCommerceEventBundle(commerceEvent: CommerceEvent): PickyBundle {
        val pickyBundle = getTransactionAttributesBundle(commerceEvent)
        var currency = commerceEvent.currency
        if (currency == null) {
            Logger.info(CURRENCY_FIELD_NOT_SET)
            currency = USD
        }
        return pickyBundle
            .putString(FirebaseAnalytics.Param.CURRENCY, currency)
            .putBundleList(FirebaseAnalytics.Param.ITEMS, getProductBundles(commerceEvent))
            .putString(FirebaseAnalytics.Event.SET_CHECKOUT_OPTION, commerceEvent.checkoutOptions)
            .putInt(FirebaseAnalytics.Event.CHECKOUT_PROGRESS, commerceEvent.checkoutStep)
    }

    fun getProductBundles(commerceEvent: CommerceEvent): Array<Bundle?> {
        val products = commerceEvent.products
        if (products != null) {
            val bundles = arrayOfNulls<Bundle>(products.size)
            var i = 0
            for (product in products) {
                val bundle = getBundle(product)
                    .putString(FirebaseAnalytics.Param.CURRENCY, commerceEvent.currency)
                bundles[i] = bundle.bundle
                i++
            }
            return bundles
        }
        return arrayOfNulls(0)
    }

    fun getTransactionAttributesBundle(commerceEvent: CommerceEvent): PickyBundle {
        val pickyBundle = PickyBundle()
        val transactionAttributes = commerceEvent.transactionAttributes
        return if (commerceEvent.transactionAttributes == null) {
            pickyBundle
        } else pickyBundle
            .putString(
                FirebaseAnalytics.Param.TRANSACTION_ID,
                transactionAttributes?.id
            )
            .putDouble(
                FirebaseAnalytics.Param.VALUE,
                transactionAttributes?.revenue
            )
            .putDouble(
                FirebaseAnalytics.Param.TAX,
                transactionAttributes?.tax
            )
            .putDouble(
                FirebaseAnalytics.Param.SHIPPING,
                transactionAttributes?.shipping
            )
            .putString(
                FirebaseAnalytics.Param.COUPON,
                transactionAttributes?.couponCode
            )
    }

    fun getBundle(product: Product): PickyBundle {
        return PickyBundle()
            .putLong(FirebaseAnalytics.Param.QUANTITY, product.quantity.toLong())
            .putString(FirebaseAnalytics.Param.ITEM_ID, product.sku)
            .putString(FirebaseAnalytics.Param.ITEM_NAME, product.name)
            .putString(FirebaseAnalytics.Param.ITEM_CATEGORY, product.category)
            .putDouble(FirebaseAnalytics.Param.PRICE, product.unitPrice)
    }

    private fun getValue(commerceEvent: CommerceEvent): Double? {
        var value = 0.0
        val products = commerceEvent.products ?: return null
        for (product in products) {
            value += product.quantity * product.unitPrice
        }
        return value
    }

    override fun onIncrementUserAttribute(
        key: String,
        incrementedBy: Number,
        value: String,
        filteredMParticleUser: FilteredMParticleUser
    ) {
        standardizeName(key, false)?.let {
            FirebaseAnalytics.getInstance(context).setUserProperty(
                it, value
            )
        }
    }

    override fun onRemoveUserAttribute(key: String, filteredMParticleUser: FilteredMParticleUser) {
        standardizeName(key, false)?.let {
            FirebaseAnalytics.getInstance(context).setUserProperty(
                it, null
            )
        }
    }

    /**
     * We are going to ignore Lists here, since Firebase only supports String "user property" values
     */
    override fun onSetUserAttribute(
        key: String,
        value: Any,
        filteredMParticleUser: FilteredMParticleUser
    ) {
        if (value is String) {
            standardizeName(key, false)?.let {
                FirebaseAnalytics.getInstance(context).setUserProperty(
                    it, standardizeValue(value, false)
                )
            }
        }
    }

    override fun onSetUserTag(s: String, filteredMParticleUser: FilteredMParticleUser) {}
    override fun onSetUserAttributeList(
        s: String,
        list: List<String>,
        filteredMParticleUser: FilteredMParticleUser
    ) {
    }

    override fun onSetAllUserAttributes(
        userAttributes: Map<String, String>,
        userAttributeLists: Map<String, List<String>>,
        filteredMParticleUser: FilteredMParticleUser
    ) {
        var userAttributes: Map<String, String>? = userAttributes
        userAttributes = standardizeAttributes(userAttributes, false)
        if (userAttributes != null) {
            for ((key, value) in userAttributes) {
                FirebaseAnalytics.getInstance(context).setUserProperty(key, value)
            }
        }
    }

    override fun supportsAttributeLists(): Boolean {
        return false
    }

    override fun onConsentStateUpdated(
        consentState: ConsentState,
        consentState1: ConsentState,
        filteredMParticleUser: FilteredMParticleUser
    ) {
    }

    fun standardizeAttributes(
        attributes: Map<String, String>?,
        event: Boolean
    ): Map<String, String>? {
        if (attributes == null) {
            return null
        }
        val attributeCopy = HashMap<String, String>()
        for ((key, value) in attributes) {
            attributeCopy[standardizeName(key, event)!!] = standardizeValue(value, event)
        }
        return attributeCopy
    }

    fun standardizeValue(valueIn: String?, event: Boolean): String {
        var value = valueIn ?: return ""
        if (event) {
            if (value.length > eventValMaxLength) {
                value = value.substring(0, eventValMaxLength)
            }
        } else {
            if (value.length > userAttributeValMaxLength) {
                value = value.substring(0, userAttributeValMaxLength)
            }
        }
        return value
    }

    fun standardizeName(nameIn: String?, event: Boolean): String? {
        var name = nameIn ?: return null
        name = name.replace("[^a-zA-Z0-9_\\s]".toRegex(), " ")
        name = name.replace("[\\s]+".toRegex(), "_")
        for (forbiddenPrefix in forbiddenPrefixes) {
            if (name.startsWith(forbiddenPrefix)) {
                name = name.replaceFirst(forbiddenPrefix.toRegex(), "")
            }
        }
        while (name.isNotEmpty() && !Character.isLetter(name.toCharArray()[0])) {
            name = name.substring(1)
        }
        if (event) {
            if (name.length > eventMaxLength) {
                name = name.substring(0, eventMaxLength)
            }
        } else {
            if (name.length > userAttributeMaxLength) {
                name = name.substring(0, userAttributeMaxLength)
            }
        }
        return name
    }

    class PickyBundle {
        val bundle = Bundle()
        fun putString(key: String?, value: String?): PickyBundle {
            if (value != null) {
                bundle.putString(key, value)
            }
            return this
        }

        fun putDouble(key: String?, value: Double?): PickyBundle {
            if (value != null) {
                bundle.putDouble(key, value)
            }
            return this
        }

        fun putLong(key: String?, value: Long?): PickyBundle {
            if (value != null) {
                bundle.putLong(key, value)
            }
            return this
        }

        fun putInt(key: String?, value: Int?): PickyBundle {
            if (value != null) {
                bundle.putInt(key, value)
            }
            return this
        }

        fun putBundleList(key: String?, value: Array<Bundle?>?): PickyBundle {
            if (value != null) {
                bundle.putParcelableArray(key, value)
            }
            return this
        }
    }

    companion object {
        const val USER_ID_FIELD_KEY = "userIdField"
        const val USER_ID_CUSTOMER_ID_VALUE = "customerId"
        const val USER_ID_EMAIL_VALUE = "email"
        const val USER_ID_MPID_VALUE = "mpid"
        private val forbiddenPrefixes = arrayOf("google_", "firebase_", "ga_")
        private const val CURRENCY_FIELD_NOT_SET = "Currency field required by Firebase was not set, defaulting to 'USD'"
        private const val USD = "USD"
        private const val eventMaxLength = 40
        private const val userAttributeMaxLength = 24
        private const val eventValMaxLength = 100
        private const val userAttributeValMaxLength = 36
        private const val KIT_NAME = "Google Analytics for Firebase"
    }
}